### DISCUSSION TITLE
dhcproxy config

### DISCUSSION THREAD

**Discussion by tanshihaj** [ipxe/ipxe#873] [Q&A]
I'm using iPXE (version `1.21.1+git-20220113.fbbdc3926-0ubuntu1` from Ubuntu Jammy repo) and faced with strange behavior which I do not understand.

In my setup there is DHCP server (10.1.0.1) and separate PXE server (10.2.0.1). DHCP server works as DHCP relay client and proxy request to PXE server. My booting sequence is following:
1. I'm booting PXE compatible firmware from network card ROM which normally detects PXE server and send proxyDHCP request to 10.2.0.1, download `ipxe.efi` from 10.2.0.1 over TFTP and boot iPXE.
2. iPXE boots up and do pretty the same actions again and should get `boot.ipxe` script from PXE server to chain loading further.

But but it seems that there is an issue with iPXE on step 2: it fails to detect PXE server correctly and begin to download `ipxe.efi` from 10.2.0.1 again. And it seems iPXE do it even before getting response from PXE server right after it get response from DHCP server. **If I interrupt iPXE with ctrl+b and get shell on step 2 I see that configs `proxydhcp/filename == ipxe.efi` & `next-server == 10.16.0.1` already. Even before sending proxyDHCP request to 10.2.0.1.** I think that it is part of the problem.

It seems that this behavior depends on latency between DHCP server & PXE server. When round-trip time is about 50ms iPXE detects PXE server and load `boot.ipxe` correctly, and if round-trip time is about 140ms it stuck in infinite loop of downloading `ipxe.efi`.

How can I debug this issue? Maybe it is some well-known issue that can be easily fixed? Quick google & search amount discussions didn't help.

**Comment by NiKiZe** [comment]
You should not have 2 DHCP servers handing out next-server and filename, remove that from you main DHCP and leave it up to your proxydhcp.

You are escentially dealing with https://ipxe.org/howto/chainloading#breaking_the_infinite_loop, and I think part of the problem is that you assume that iPXE should not use your proxydhcp ?

You don't say which software you are using on the server side, you can find some documentation on above url, and a separate proxydhcp only example at https://ipxe.org/appnote/proxydhcp

**Reply by tanshihaj** [reply]
> You should not have 2 DHCP servers handing out next-server and filename, remove that from you main DHCP and leave it up to your proxydhcp.

Yes, I know, right now DHCP server (10.1.0.1) do not send any next-server of filename - only IP leases and router IP address. Everything connected with network boot handled by PXE server (10.2.0.1).

> You are escentially dealing with https://ipxe.org/howto/chainloading#breaking_the_infinite_loop, and I think part of the problem is that you assume that iPXE should not use your proxydhcp ?

Not exactly - I assume that iPXE **will always** use proxydhcp which it does not do or do depending on network latency.

> You don't say which software you are using on the server side, you can find some documentation on above url, and a separate proxydhcp only example at https://ipxe.org/appnote/proxydhcp

I'm using dnsmasq which config is pretty the same as in appnote. The issue is that it didn't work persistently.

**Reply by NiKiZe** [reply]
> Not exactly - I assume that iPXE **will always** use proxydhcp which it does not do or do depending on network latency.
> 

Almost, it should wait for proxydhcp if filename and next-server is missing, and there is no "dont wait for proxydhcp" options in the dhcp replies. and it also needs to use default dhcp process. If you for example use embedded script and only do ifopen, then the cached dhcp information will/should be reused.

Have you used tcpdump or wireshark to see what the dhcp packets actually contain?

commit fbbdc3926 is quite old. please use current master, prebuilt binaries can be found at boot.ipxe.org if you really don't want to build it yourself. (but you should really try to build it locally)

**Reply by tanshihaj** [reply]
> Almost, it should wait for proxydhcp if filename and next-server is missing, and there is no "dont wait for proxydhcp" options in the dhcp replies.

Yes, I understand it. DHCP server do not set boot file name & next server.

> and it also needs to use default dhcp process.

Yes, I understand it.

> Have you used tcpdump or wireshark to see what the dhcp packets actually contain?

Yes, I've intercepted packages from DHCP server's interface faced to the PXE & iPXE client. And DHCP server does not send boot file name & next server fields: Here is pcap:https://drive.google.com/file/d/1nFN2HVlJJdPybHo83e-GtS15k711y2_T/view?usp=share_link. First 7 packages is PXE client from ROM and last 5 - iPXE client.

> commit https://github.com/ipxe/ipxe/commit/fbbdc39260cf37aa749e897e773f59807d1b8362 is quite old. please use current master, prebuilt binaries can be found at boot.ipxe.org if you really don't want to build it yourself. (but you should really try to build it locally)

Just checked - the last version from boot.ipxe.org does the same thing.

**Reply by mcb30** [reply]
@tanshihaj The interesting data point from your pcap file is in packet 8: the allegedly first DHCPDISCOVER coming from iPXE.  In this packet, the `dhcp.secs` field has the value 16.

iPXE uses the `dhcp.secs` field to convey some diagnostic information:
https://github.com/ipxe/ipxe/blob/master/src/net/udp/dhcp.c#L1150-L1163

In this case, the value of 16 shows that:
* iPXE has not received any DHCPOFFER (i.e. with a valid client IP address) at this point
* iPXE has not received any ProxyDHCPOFFER (i.e. with option 60 set to `PXEClient`) at this point
* This is the fourth DHCPDISCOVER that iPXE has transmitted

So: iPXE has sent multiple DHCPDISCOVERs (missing from your pcap file).  It has had no responses yet.  It finally gets a response from your DHCP server.  By this time, it has already been attempting DHCPDISCOVERs for longer than is necessary to wait for a potential ProxyDHCPOFFER, and so when it finally receives a DHCPOFFER it moves straight to the DHCPREQUEST phase.  The DHCPREQUEST completes successfully when a DHCPACK response is received.

At this point, iPXE has:

* The `net0.dhcp` settings block, corresponding to the successful DHCPACK
* The `proxydhcp` settings block that remains present since iPXE started up.  (iPXE will pick up cached DHCP and ProxyDHCP packets from the previous stage bootloader.)

So: iPXE has a valid IP configuration for `net0`, and a valid boot filename (the cached `proxydhcp.filename`).  It then obediently boots from this `proxydhcp.filename`, since no filename was provided in the DHCPACK.


I hope that explains what is happening.  The root cause of the problem is that your DHCP server and relay are failing to respond to the first three DHCPDISCOVER packets that iPXE sends out.  Fix that, and the problem will go away.

**Reply by tanshihaj** [reply]
> I hope that explains what is happening.

Yes, it is really helpful, thank you! It seems there is some network issues somewhere between the NIC and the router. I will try to get pcap from the switch to which the NIC connected.

**Comment by stappersg** [comment]
On Wed, Feb 01, 2023 at 08:50:03AM -0800, Kamil Zaripov wrote:
> Here is pcap: https://drive.google.com/file/d/1nFN2HVlJJdPybHo83e-GtS15k711y2_T/view?usp=share_link.

That link does require some login that did block fetching.


> First 7 packages is PXE client from ROM and last 5 - iPXE client.
 
Yes, I would like to analyze the .pcap


> > commit https://github.com/ipxe/ipxe/commit/fbbdc39260cf37aa749e897e773f59807d1b8362
> > is quite old. please use current master, prebuilt binaries can
> > be found at boot.ipxe.org if you really don't want to build it
> > yourself.
> 
> Just checked - the last version from boot.ipxe.org does the same thing.



> > (but you should really try to build it locally)

So you can debug information.


Groeten
Geert Stappers
-- 
Silence is hard to parse
